Adding RT-3.52: Multidimensional test for Static GUE Encap/Decap based on BGP path selection and selective DSCP marking

ASHNA-AGGARWAL-KEYSIGHT opened the pull request:
Readme Location: https://github.com/openconfig/featureprofiles/blob/main/feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/README.md

Raised an issue for the test not passing completely: https://partnerissuetracker.corp.google.com/issues/441232997

Logs attached: https://partnerissuetracker.corp.google.com/issues/415458482#comment142

ASHNA-AGGARWAL-KEYSIGHT committed:
Adding RT-3.52: Multidimensional test for Static GUE Encap/Decap based on BGP path selection and selective DSCP marking

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/metadata.textproto:
```diff
@@ -4,4 +4,23 @@
+    next_hop_group_config_unsupported: true
+    policy_forwarding_unsupported: true
+    interface_policy_forwarding_unsupported: true
+    ipv4_static_route_with_ipv6_nh_unsupported: true
```
Why is this deviation needed? Isnt the test requiring Static routes for IPv6 destination pointing at a IPv4 NH?

ASHNA-AGGARWAL-KEYSIGHT committed:
added changes

ASHNA-AGGARWAL-KEYSIGHT committed:
Added changes

ASHNA-AGGARWAL-KEYSIGHT commented:
Need more clarification on the issue:  https://partnerissuetracker.corp.google.com/issues/441232997

ASHNA-AGGARWAL-KEYSIGHT committed:
Updated test according to bug 441232997

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2144 @@
+		// IPv6Layer:   innerGUEIPLayerIPv6,
+	}
+
+	decapValidation = &packetvalidationhelpers.PacketValidation{
```
Looks like we are expecting the ATE to set a TTL of 64 before forwarding the packet to the DUT? Rather, can we explicitly set ttl=64 on the OTG flow definition for Flow-Set#3 and Flow-Set#4? This imo will make the test more robust, no?

ASHNA-AGGARWAL-KEYSIGHT committed:
addressed review comments

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2115 @@
+
+	outerGUEIPLayerIPv4.Tos = uint8(dscpValue["BE1"])
+	innerGUEIPLayerIPv4.Tos = uint8(dscpValue["BE1"])
+	if err := validatePacket(t, ate, encapValidation); err != nil {
```
According to the README, ATE2_INTERNAL_TE11.v4/30 is advertised by ATE_Port3. I assumed that ATE2_INTERNAL_TE11.v4/32 is advertised by ATE_Port2, which is why we are receiving encapsulated packets on Port 2. I have therefore included the necessary validation.

To resolve this issue, I modified the configuration so that ATE_Port3 now advertises /32. This change should help us achieve the desired outcome, as you mentioned in your comment. Could you please confirm whether this trial configuration is correct? If it is, we will need to make a minor update to the README, and I will update the PR.

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2115 @@
+				EthFlow:       &otgconfighelpers.EthFlowParams{SrcMAC: otgBGPConfig[1].otgPortData[1].MAC, DstMAC: port2DstMac},
+				VLANFlow:      &otgconfighelpers.VLANFlowParams{VLANId: otgBGPConfig[1].otgPortData[0].Subinterface},
+				IPv4Flow:      &otgconfighelpers.IPv4FlowParams{IPv4Src: dutloopback0.IPv4, IPv4Dst: bgpInternalTE11.IPv4, TTL: ttl},
+				UDPFlow:       &otgconfighelpers.UDPFlowParams{UDPDstPort: udpEncapPort},
```
On Port 1, we expect to receive decapsulated packets to verify whether the decapsulation process is working correctly. I will send an encapsulated packet from Port 2/Port 3 to Port 1. The DUT will then decapsulate the packet, allowing us to confirm the results on Port 1. The packet sent to Port 1 is similar to the packet we received after encapsulation on Port 2/Port 3.

Please let me know if my understanding is correct.

ASHNA-AGGARWAL-KEYSIGHT committed:
updated code

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2024 @@
+package staticgueencapbgppathselection_test
```
Lets wait for https://github.com/openconfig/featureprofiles/pull/4927 to be approved. We will need to modify the code to change to IPv6 Transport

ksgireesha commented:
@ASHNA-AGGARWAL-KEYSIGHT  - PR #4927 is merged. Keysight to replace v4 tunnels with v6 equivalents as detailed in the Description of the bug b/471388109

ASHNA-AGGARWAL-KEYSIGHT committed:
updated code with latest readme changes

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
updated branch

ASHNA-AGGARWAL-KEYSIGHT committed:
fixed the util test issue

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
addressed review comments

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
fixed issues

ASHNA-AGGARWAL-KEYSIGHT committed:
added fix

ASHNA-AGGARWAL-KEYSIGHT committed:
Revert "addressed review comments"

This reverts commit 76252041420b2654cd9ca8b4d3a83d10753c4bed.

ASHNA-AGGARWAL-KEYSIGHT committed:
fixing the issues

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2013 @@
+				FlowName:      "flowSet5-v6-1",
+				EthFlow:       &otgconfighelpers.EthFlowParams{SrcMAC: otgBGPConfig[1].otgPortData[1].MAC, DstMAC: port2DstMac},
+				VLANFlow:      &otgconfighelpers.VLANFlowParams{VLANId: otgBGPConfig[1].otgPortData[1].Subinterface},
+				IPv6Flow:      &otgconfighelpers.IPv6FlowParams{IPv6Src: dutloopback0.IPv6, IPv6Dst: bgpInternalTE11.IPv6, HopLimit: ttl},
```
The README specifies that Flow-Set #5 should be sent Unencapsulated (Line 184 in README). However, the code here adds IPv6Flow and UDPFlow headers, making it GUE encapsulated. Please verify if this should be unencapsulated as per the README.

ASHNA-AGGARWAL-KEYSIGHT replied:
@sachendras For flowSwt-5, the traffic is directed from ATE-port2 to ATE-port1. A decapsulation tunnel has been configured on $ATE2_INTERNAL_TE10.v6 & TE11, which performs the decapsulation of the packets. As a result, ATE-port1 receives the unencapsulated packets. This setup ensures compliance with the validation requirement: “Unencapsulated flows from ATE2 to ATE1_Port1 must achieve 100% success (zero packet loss).”

Please let me know if my understanding is correct; if not, I will make the necessary changes.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2013 @@
+	advertiseRoutesWithiBGP([]string{ate2InternalPrefixesV4List[0]}, ate2ppnh1.IPv6, true, fmt.Sprintf("%s-CBGP-1", atePort2RoutesV4))
+	advertiseRoutesWithiBGP([]string{ate2InternalPrefixesV6List[0]}, ate2ppnh1.IPv6, false, fmt.Sprintf("%s-CBGP-1", atePort2RoutesV6))
+
+	flowsets := []string{"flowSet1", "flowSet5"}
```
The README for RT-3.52.2 states that flows in Flow-Set #1, Flow-Set #2, and Flow-Set #5 should be active. The code here only includes "flowSet1" and "flowSet5". Please check if "flowSet2" should also be included.

ASHNA-AGGARWAL-KEYSIGHT replied:
As mentioned in the readme . $ATE2_C.IBGP.v6 advertises $ATE2_INTERNAL[6-8].v[46] Please Note: These prefixes are gradually advertised by $ATE2_C.IBGP.v6 in different Sub tests, so accordingly i am making the flows active and checking the traffic. From RT-3.52.5 onwards i  added flowset2.

Let me know if it is fine

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2013 @@
+	otgConfig.Flows().Clear()
+
+	for _, flow := range flowsets {
+		otgConfig.Flows().Append(flowGroups[flow].Flows[0:2]...)
```
The code here only appends the first two flows of the selected flowsets (Flows[0:2]). The README implies that all flows in the specified sets should be active to test traffic migration realistically. Please verify if all flows should be enabled.

sachendras commented on internal/iputil/iputil.go:
```diff
@@ -248,3 +249,45 @@ func IncrementMAC(startMAC string, i int) (string, error) {
+		return nil, fmt.Errorf("expected IPv6 mask, got %d bits", bits)
+	}
+
+	firstIP := binary.BigEndian.Uint64(netCIDR.IP)
```
binary.BigEndian.Uint64 reads only the first 8 bytes of the 16-byte IPv6 address/mask. This truncates the IPv6 address to 64 bits and ignores the lower half of the address.

sachendras commented on internal/iputil/iputil.go:
```diff
@@ -248,3 +249,45 @@ func IncrementMAC(startMAC string, i int) (string, error) {
+
+	firstIP := binary.BigEndian.Uint64(netCIDR.IP)
+	netMask := binary.BigEndian.Uint64(netCIDR.Mask)
+	lastIP := (firstIP & netMask) | (netMask ^ 0xffffffff)
```
The use of 0xffffffff (a 32-bit literal) here causes incorrect calculation when used with netMask (which is uint64). This appears to be a remnant from the IPv4 version of this function (GenerateIPs on line 40).

sachendras commented on internal/iputil/iputil.go:
```diff
@@ -248,3 +249,45 @@ func IncrementMAC(startMAC string, i int) (string, error) {
+
+	for i := firstIP; i <= lastIP; i++ {
+		ipv6 := make(net.IP, 16)
+		binary.BigEndian.PutUint64(ipv6, i)
```
binary.BigEndian.PutUint64 only fills the first 8 bytes of the 16-byte ipv6 slice. This means you are only varying the upper 64 bits (network prefix) and leaving the lower 64 bits as zero (plus the increment in line 292). If the intention was to generate hosts within a prefix, this will not work as expected. Consider using GenerateIPv6s (line 224) which uses big.Int and correctly handles 128-bit addresses.

ASHNA-AGGARWAL-KEYSIGHT committed:
updated code

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
fixed conflicts

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2013 @@
+	}
+}
+
+func testTrafficMigration(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, otgConfig gosnappi.Config, flowNames map[string][]string, dscpVal string) {
```
Can you please recheck this function and ensure all the flows as defined in RT-3.52.2 to RT-3.52.10 are covered? At the moment the code seems to consider only flowset# 1 and 5

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2013 @@
+	testTrafficMigration(t, dut, ate, otgConfig, flowNames, "AF4")
+}
+
+func testDUTDecapNode(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, otgConfig gosnappi.Config) {
```
Please also include validatePacket to the end of this function to align with the readme. This needs verified for flowset# 3 and 4 for RT-3.52.7 onwards where the DUT is working as a decap node.

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
Addresed review comments

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

jjw-a-0617 commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2053 @@
+		bgpCreateNbr(t, dutAS, dut, cfg.bgpCfg)
+	}
+
+	configureLoopback(t, dut)
```
`bgpCreateNbr(...)` is called inside the `for _, cfg := range otgBGPConfig` loop, and `configureLoopback(t, dut)` is called after it. Two consequences:

1. `otgBGPConfig[*].bgpCfg[*].srcIp` is a package-init-time copy of the hardcoded `dutloopback0` (203.0.113.10 / 2001:db8:100::10).
2. `configureLoopback`'s else branch updates `dutloopback0` to match the DUT's actual Loopback0, but only after BGP has already been pushed with the stale values.

On a DUT whose Loopback0 isn't the hardcoded pair, the DUT's BGP router-id and every neighbor's transport/local-address get set to an address not configured anywhere on the box. TCP can't bind → no peering → 0 prefixes → 100% loss → all 10 subtests fail in the first BGP-up check.

ASHNA-AGGARWAL-KEYSIGHT replied:
We have set the value of src-ip to the the loopback address which is being configured on the DUT. Once all the configurations are done from DUT and OTG side, we are starting the protocols and checking the whether peers are established or not.

jjw-a-0617 replied:
Thanks for replying!

```
func configureLoopback(t *testing.T, dut *ondatra.DUTDevice) {
	// Configure interface loopback
	loopbackIntfName = netutil.LoopbackInterface(t, dut, 0)
	lo0 := gnmi.OC().Interface(loopbackIntfName).Subinterface(0)
	ipv4Addrs := gnmi.LookupAll(t, dut, lo0.Ipv4().AddressAny().State())
	ipv6Addrs := gnmi.LookupAll(t, dut, lo0.Ipv6().AddressAny().State())
	if len(ipv4Addrs) == 0 && len(ipv6Addrs) == 0 {
		loop1 := dutloopback0.NewOCInterface(loopbackIntfName, dut)
		loop1.Type = oc.IETFInterfaces_InterfaceType_softwareLoopback
		gnmi.Update(t, dut, gnmi.OC().Interface(loopbackIntfName).Config(), loop1)
	} else {
		v4, ok := ipv4Addrs[0].Val()
		if ok {
			dutloopback0.IPv4 = v4.GetIp()
		}
		v6, ok := ipv6Addrs[0].Val()
		if ok {
			dutloopback0.IPv6 = v6.GetIp()
		}
		t.Logf("Got DUT IPv4 loopback address: %v", dutloopback0.IPv4)
		t.Logf("Got DUT IPv6 loopback address: %v", dutloopback0.IPv6)
	}
}
```

The issue is that `dutloopback0` can be overriden when taking the else branch, but the bgp session is configured based on its previous hardcoded address.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2053 @@
+	cfgplugins.VerifyDUTBGPEstablished(t, dut)
+
+	// Validating no prefixes are exchanged over the IBGP peering between $ATE2_C.IBGP.v6 and $DUT_lo0.v6
+	validatePrefixes(t, dut, otgBGPConfig[1].otgPortData[2].IPv6, false, 0, 0)
```
Should this be the following instead?
## Corrected?: 

```
validatePrefixes(t, dut, otgBGPConfig[1].otgPortData[1].IPv6, false, 0, 0) 
```

Explanation:

README Requirement: Ensure no prefixes are exchanged over the IBGP peering between $ATE2_C.IBGP.v6 (VLAN 200 peer) and $DUT_lo0.v6.

Here, seems otgBGPConfig[1].otgPortData[2].IPv6 evaluates to the IPv6 of R300 (VLAN 300 peer, which is the Multipath $ATE2_M.IBGP.v6 peer). Isn't that the case?

sachendras commented on internal/otg_helpers/packetvalidationhelpers/packetvalidationhelpers.go:
There seems to be a bug here:
README Requirement: "During this decapsulation process, the system must not transfer the DSCP and TTL bits from the outer header to the inner header."

**Bug:**
Since packetVal.IPv4Layer.Tos is 0, no validation is done on the ToS byte of the decapsulated packet received on Port 1.
Impact: If the DUT incorrectly transfers/overwrites the inner packet's DSCP (e.g. making it match the outer GUE header's DSCP), the test will still pass because the validation is skipped. 




Can you please replace the following section with below?

## Corrected Code:

```
for packet := range packetSource.Packets() {
 		if ipLayer := packet.Layer(layers.LayerTypeIPv4); ipLayer != nil {
 			ip, _ := ipLayer.(*layers.IPv4)
			// If DstIP is specified, skip packets that don't match it
			if packetVal.IPv4Layer.DstIP != "" && ip.DstIP.String() != packetVal.IPv4Layer.DstIP {
				continue 
			}
 			if !packetVal.IPv4Layer.SkipProtocolCheck {
 				if uint32(ip.Protocol) != packetVal.IPv4Layer.Protocol {
```

Please also update testDecapTraffic in the test file to validate each flow's DSCP using the new IP-filtering capability:

## Corrected code:

```
func testDecapTraffic(t *testing.T, dut *ondatra.DUTDevice, ate *ondatra.ATEDevice, otgConfig gosnappi.Config) {
	t.Log("Validate GUE Decapsulation with v4 traffic")

	// Map of destination IP to expected shifted DSCP (TOS)
	v4DecapExpectations := map[string]uint32{
		ate1UserPrefixesV4List[0]: expectedDscpValue["BE1"], // 0
		ate1UserPrefixesV4List[1]: expectedDscpValue["AF1"], // 40
		ate1UserPrefixesV4List[2]: expectedDscpValue["AF2"], // 72
		ate1UserPrefixesV4List[3]: expectedDscpValue["AF3"], // 104
		ate1UserPrefixesV4List[4]: expectedDscpValue["AF4"], // 136
	}

	for dstIP, expectedTos := range v4DecapExpectations {
		decapInner := *innerGUEIPLayerIPv4
		decapInner.SkipProtocolCheck = true
		decapInner.Tos = uint8(expectedTos)
		decapInner.DstIP = dstIP // Trigger the new DstIP filtering in helper
		
		decapValidation.IPv4Layer = &decapInner
		if err := validatePacket(t, ate, decapValidation); err != nil {
			t.Errorf("DSCP preservation validation failed for DstIP %s (expected TOS %d): %v", dstIP, expectedTos, err)
		}
	}
}
```


Repeat similar logic for IPv6 decap validation using expectedDscpValue/dscpValue)

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2053 @@
+
+	outerGUEIPLayerIPv6 = &packetvalidationhelpers.IPv6Layer{
+		NextHeader: 17,
+		HopLimit:   ttl,
```
Readme says: "The outer header TTL should be 63 upon arrival at ATE2_Port1 (before decapsulation). Please check this on the ATE2."

However the code checks for Hoplimit = TTL which is 64, right?

## Corrected code:

```
HopLimit:   ttl - 1,
```

ASHNA-AGGARWAL-KEYSIGHT replied:
@sachendras The GUE encapsulated packet will be:

Ethernet → Outer IP → UDP → GUE → Inner IP.
Currently, in the inner IP layer, I see the TTL being decremented by 1. Could you please confirm if my understanding is correct? You are not referring to the TTL of the outer IP after the Ethernet packet, but rather to the inner IP TTL being decremented.

sachendras commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2053 @@
+
+	outerv6GUEIPLayerIPv6 = &packetvalidationhelpers.IPv6Layer{
+		NextHeader: 17,
+		HopLimit:   ttl,
```
Same here,

Readme says: "The outer header TTL should be 63 upon arrival at ATE2_Port1 (before decapsulation). Please check this on the ATE2."

However the code checks for Hoplimit = TTL which is 64, right?

## Corrected code:

```
HopLimit:   ttl - 1,
```

ASHNA-AGGARWAL-KEYSIGHT committed:
addressed review comments

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
fixed issues

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ksgireesha commented:
Hi @rohit-rp , we have LGTM from @sachendras. Please review, validate and merge the PR

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
updated testregistry

navaneethyv commented:
/gemini review

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2077 @@
+	// Validating no prefixes are exchanged over the IBGP peering between $ATE2_C.IBGP.v6 and $DUT_lo0.v6
+	validatePrefixes(t, dut, otgBGPConfig[1].otgPortData[1].IPv6, false, 0, 0)
+
+	sendTrafficCapture(t, ate, []string{"all"})
```
After starting protocols we are validating BGP established on both DUT and ATE side, still do we really need WaitForARP check?

navaneethyv commented:
Hi @ASHNA-AGGARWAL-KEYSIGHT  requesting you to fix the Gemini review comments and add response comments to it once done and resolve it. Thanks. Also the branch has conflicts that needs to be fixed.

ASHNA-AGGARWAL-KEYSIGHT committed:
addressed review comments

ASHNA-AGGARWAL-KEYSIGHT commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -895,12 +896,13 @@ func aristaGueDecapCLIConfig(t *testing.T, dut *ondatra.DUTDevice, params OcPoli
 							tunnel decap-ip %s
-							tunnel decap-interface %s
-							`, params.GUEPort, decapProto, params.IPType, params.GUEPort, params.AppliedPolicyName, params.TunnelIP, params.InterfaceID)
+							`, params.GUEPort, decapProto, params.IPType, params.TunnelIP)
```
added code

ASHNA-AGGARWAL-KEYSIGHT commented on internal/cfgplugins/policyforwarding.go:
```diff
@@ -920,6 +922,38 @@ func aristaGreDecapCLIConfig(t *testing.T, dut *ondatra.DUTDevice, params OcPoli
+				cli := fmt.Sprintf(`
+				policy-map type quality-of-service %s
+   					class class-default
+      				set dscp cs%d
```
added

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2077 @@
+	if !ok {
+		t.Errorf("no sent prefixes found")
+	}
+
```
added code

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2077 @@
+	cfgplugins.PushDUTHardwareInitConfig(t, dut, hardwareInitCfg)
+}
+
+func TestStaticGue(t *testing.T) {
```
added cleanup for static routes, policy forwarding configurations, and decap groups

ASHNA-AGGARWAL-KEYSIGHT commented on feature/policy_forwarding/encapsulation/otg_tests/staticgueencap_and_bgp_path_selection/staticgueencap_and_bgp_path_selection_test.go:
```diff
@@ -0,0 +1,2077 @@
+		txPkts := gnmi.Get(t, otg, gnmi.OTG().Flow(flow.Name()).Counters().OutPkts().State())
+
+		totalTxFromATE += txPkts
+	}
```
added code

ASHNA-AGGARWAL-KEYSIGHT commented on internal/iputil/iputil.go:
```diff
@@ -262,6 +262,54 @@ func IPEqual(got, want string) bool {
+		nextInt.Mod(nextInt, pmax)
+		ipv6 := nextInt.FillBytes(make([]byte, 16))
+		p, _ := netip.ParsePrefix(fmt.Sprintf("%v/%d", net.IP(ipv6), maskSize))
+		entries = append(entries, p.Addr().String())
```
added code

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ASHNA-AGGARWAL-KEYSIGHT committed:
fixed issues

ASHNA-AGGARWAL-KEYSIGHT commented:
> Hi @ASHNA-AGGARWAL-KEYSIGHT requesting you to fix the Gemini review comments and add response comments to it once done and resolve it. Thanks. Also the branch has conflicts that needs to be fixed.

Hi @navaneethyv I have addressed the review comments.

ASHNA-AGGARWAL-KEYSIGHT committed:
Merge branch 'main' into fpt_otg_RT_3_52

ksgireesha commented:
> > Hi @ASHNA-AGGARWAL-KEYSIGHT requesting you to fix the Gemini review comments and add response comments to it once done and resolve it. Thanks. Also the branch has conflicts that needs to be fixed.
> 
> Hi @navaneethyv I have addressed the review comments.

@rohit-rp - Keysight addressed all the comments here and we have LGTM from @sachendras. Please review, validate, approve and merge this PR.
